### ISSUE TITLE
Envoy SDS custom validator issue

### ISSUE THREAD

**Issue by StupidScience** [spiffe/spire#3001]
Hello.

Some context for beginning. We use PKI based federation with root CA and intermediate ones for trust domains. Currently we're upgrading our spire stack from 1.0.1 to 1.1.4 and faced situation when suddenly our federation implementation are not working properly for cross domain communications due to change introduced in #2435 and available from version 1.0.2. It's kinda breaking change for us and blocking our upgrade or force to do some adventurous stuff.

Can we introduce some option for SDS like `disable_custom_validation` to fallback to the old behaviour and disable Envoy SDS SPIFFE Custom validator for any version of envoy? Moreover we'll need to backport it to 1.1. We're going to migrate our federation setup but it can be difficult without this option.

I can send PR for this.

* **Version**: 1.0.2+
* **Platform**: any
* **Subsystem**: agent

**Comment by azdagron** [comment]
Hey @StupidScience, can you elaborate on exactly what has broken in your federation setup and maybe share how you configure the validation_context in Envoy?

**Comment by evan2645** [comment]
More information would be great.

> We use PKI based federation with root CA and intermediate ones for trust domains.

There are a lot of downsides to this approach, but it's probably not the place to go into that.

If you have access to a test environment, you may be able to work around this problem by manually setting federated bundles for the various trust domain names, setting the shared bundle for each one. I'm not quite sure how JWT-SVID will behave in this scenario, certainly it won't work as expected, I think it depends on your exact deployment topology (e.g. nested SPIRE or not, where is the upstream, etc).

> It's kinda breaking change for us

Very sorry about this, it's something we try to be very careful about. We'll make sure you're supported in one way or another

**Comment by StupidScience** [comment]
So, I will try to explain better.

Our validation context on 1.0.1 looks like these:
```json
{
    "@type": "type.googleapis.com/envoy.admin.v3.SecretsConfigDump",
    "dynamic_active_secrets": [
        ...
        {
            "name": "spiffe://first-trust-domain",
            "secret": {
                "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret",
                "name": "spiffe://first-trust-domain",
                "validation_context": {
                    "trusted_ca": {
                        "inline_bytes": "root-ca-cert-base64"
                    }
                }
            }
        }
    ]
}
```

And after spire agent upgrade we should have something like (can't check for sure atm):
```json
{
    "@type": "type.googleapis.com/envoy.admin.v3.SecretsConfigDump",
    "dynamic_active_secrets": [
        ...
        {
            "name": "spiffe://first-trust-domain",
            "secret": {
                "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret",
                "name": "spiffe://first-trust-domain",
                "validation_context": {
                    "custom_validator_config": {
                        "name": "envoy.tls.cert_validator.spiffe",
                        "typed_config": {
                            "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.SPIFFECertValidatorConfig",
                            "trust_domains": {
                                "name": "spiffe://first-trust-domain",
                                "trust_bundle": {
                                    "inline_bytes": "root-ca-cert-base64"
                                }
                            }
                        }
                    }
                }
            }
        }
    ]
}
```

The problem is that envoy's SPIFFE Certificate Validator won't check if certificate is valid or not since Envoy selects the trust bundle according to the presented SAN before validate the certificate because SVID will be signed by another (let's say second-trust-domain) trust domain CA. But in terms of PKI they're both issued by root ca so should be valid and that's how it work for us now.

> There are a lot of downsides to this approach, but it's probably not the place to go into that.

Maybe, but I think that our federation implementation was created when there was no any proposed by spire so may be that was the only available option. Anyway we'll look for migration paths.

**Comment by azdagron** [comment]
I see. One option would be to set the federation bundle for each trust domain on each SPIRE server cluster and then update registration entries to include the federated trust domains in the FederatesWith list. If you move to an official federation mechanism.

**Comment by azdagron** [comment]
I'm also not super opposed to providing a flag to return the old validation context configuration but if there are other viable workarounds I think it's worth exploring those.

**Comment by evan2645** [comment]
> Maybe, but I think that our federation implementation was created when there was no any proposed by spire so may be that was the only available option. Anyway we'll look for migration paths.

Yes, I understand .. I don't think you are alone in this configuration :)

**Comment by azdagron** [comment]
Hey @StupidScience, the maintainers got together and came to a consensus that the proposed change you describe would be welcome and something we'd be willing to backport to previous releases since it breaks the upgrade path for deployments like yours.

I've your still willing to put up a PR, we'd be happy to help get it merged.

**Comment by StupidScience** [comment]
@azdagron great news, thank you. I've sent corresponding PR https://github.com/spiffe/spire/pull/3009
